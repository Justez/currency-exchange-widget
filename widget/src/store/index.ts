import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';

import createRootReducer from './root-reducer';
import sagas from './root-saga';

const configureStore = () => {
  const sagaMiddleware = createSagaMiddleware();

  const middlewares = [
    sagaMiddleware,
  ];

  const instance = createStore(
    createRootReducer(),
    composeWithDevTools(applyMiddleware(...middlewares)),
  );

  sagaMiddleware.run(sagas);

  return instance;
};

export type State = ReturnType<typeof store.getState>;

const store = configureStore();

export default store;
