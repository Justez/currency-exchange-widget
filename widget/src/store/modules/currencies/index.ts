//@ts-ignore
import { createActions } from 'redux-actions';

export const NAMESPACE = 'currencies';

export const actions = createActions(
  'SET_SELECTED_CURRENCY',
  'FLIP_SELECTED_CURRENCIES',
  { prefix: NAMESPACE },
);
