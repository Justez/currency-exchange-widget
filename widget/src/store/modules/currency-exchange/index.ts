//@ts-ignore
import { createActions } from 'redux-actions';

export const NAMESPACE = 'currencyRates';

export const actions = createActions(
  'START_POLLING_CURRENCY_RATES',
  'STOP_POLLING_CURRENCY_RATES',
  'GET_CURRENCY_RATES',
  'GET_CURRENCY_RATES_REQUEST',
  'GET_CURRENCY_RATES_SUCCESS',
  'GET_CURRENCY_RATES_FAILURE',
  'SET_CURRENCY_RATES',
  { prefix: NAMESPACE },
);
