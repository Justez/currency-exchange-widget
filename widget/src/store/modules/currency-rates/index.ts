import { createActions } from 'redux-actions';

export const NAMESPACE = 'currencyRates';

export const actions = createActions(
  'FLIP_RATES',
  'GET_CURRENCY_RATES',
  'GET_CURRENCY_RATES_REQUEST',
  'GET_CURRENCY_RATES_SUCCESS',
  'GET_CURRENCY_RATES_FAILURE',
  'SET_CURRENCY_RATES',
  'RESET_RATES',
  { prefix: NAMESPACE },
);
