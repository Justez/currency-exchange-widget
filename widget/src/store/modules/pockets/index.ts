import { createActions } from 'redux-actions';

export const NAMESPACE = 'pockets';

export const actions = createActions(
  'PLACE_SUM',
  'PLACE_ALL',
  'PLACE_EXCHANGE',
  'PLACE_EXCHANGE_REQUEST',
  'PLACE_EXCHANGE_SUCCESS',
  'PLACE_EXCHANGE_FAILURE',
  'SET_PLACED_AMOUNT',
  { prefix: NAMESPACE },
);
