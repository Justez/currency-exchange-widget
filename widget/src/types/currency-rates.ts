import { CurrencyMap } from 'types';

export type CurrencyRateResponse = {
  data: {
    base: string;
    rates: {
      [key in CurrencyMap]: number;
    };
  }
}

export type CurrencyRateError = {
    error: boolean;
    status: number;
    message: string;
    description: string;
}