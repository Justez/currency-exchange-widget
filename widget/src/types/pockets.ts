import { CurrencyMap } from './'

export type Pocket = {
    currency: CurrencyMap,
    sum: string;
    placedSum: string;
}

export type Pockets = Array<Pocket>