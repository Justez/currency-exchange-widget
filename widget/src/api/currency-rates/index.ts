import axios from 'axios'

import { Currencies,CurrencyRateResponse, CurrencyRateError } from 'types';

const URL = 'https://openexchangerates.org/api/latest.json';

export const getCurrencyRates = (obj: Currencies) => axios.get(URL, {
    params: {
        app_id: '2b3e9170d55c46fead589f4fdeac1fb8',
        // base: obj.out.toUpperCase(),
    }
})
    .then((res: CurrencyRateResponse) => res.data)
    .catch((err: CurrencyRateError) => err);