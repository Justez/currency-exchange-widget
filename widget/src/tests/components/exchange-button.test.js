import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';

import store from 'store';
import { actions as pocketsActions } from 'store/modules/pockets';
import { actions as currencyRatesActions } from 'store/modules/currency-rates';
import ExchangeButton from 'components/exchange-button';

const component = (
    <Provider store={store}>
        <ExchangeButton />
    </Provider>
);

let wrapper;
let button;

describe('Exchange button', () => {
    const load = () => {
        wrapper = mount(component);
        button = wrapper.find('button');
    };

    beforeAll(() =>  load());

    it('is rendered', () => {
        expect(button.exists()).toBeTruthy();
    });

    it('is disabled on init state', () => {
        expect(button.prop('disabled')).toBeTruthy();
    });

    it('is enabled if rates are present', () => {
        store.dispatch(currencyRatesActions.setCurrencyRates({ rate: 1.22, reverse: 0.88 }));
        load();
        expect(button.prop('disabled')).toBeFalsy();
    });

    it('is disabled exchange submitted', () => {
        store.dispatch(pocketsActions.placeExchangeRequest());
        load();
        expect(button.prop('disabled')).toBeTruthy();
    });

    it('is enabled again after exchange submitted', () => {
        store.dispatch(pocketsActions.placeExchangeSuccess());
        load();
        expect(button.prop('disabled')).toBeFalsy();
    });

    it('is disabled when loading rates', () => {
        store.dispatch(currencyRatesActions.getCurrencyRates());
        load();
        expect(button.prop('disabled')).toBeTruthy();
    });

    it('is enabled again after loading rates', () => {
        store.dispatch(currencyRatesActions.getCurrencyRatesSuccess());
        load();
        expect(button.prop('disabled')).toBeFalsy();
    });

});

describe('ExchangeButton snapshots', () => {
    const render = () => renderer.create(component).toJSON();

    it('renders enabled', () => {
        expect(render()).toMatchSnapshot();
    });

    it('renders disabled after store change', () => {
        act(() => {
            store.dispatch(currencyRatesActions.getCurrencyRates());
        });
        expect(render()).toMatchSnapshot();
    });
});