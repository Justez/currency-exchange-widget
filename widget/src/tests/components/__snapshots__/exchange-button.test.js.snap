// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeButton snapshots renders disabled after store change 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary Mui-disabled MuiButton-fullWidth Mui-disabled"
  disabled={true}
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={-1}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    Exchange
  </span>
</button>
`;

exports[`ExchangeButton snapshots renders enabled 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-fullWidth"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    Exchange
  </span>
</button>
`;
